ownCloud/Client/PhotoSelectionViewController.swift

```diff
@@ -0,0 +1,337 @@
+	required init?(coder aDecoder: NSCoder) {
+		fatalError("not implemented")
+	}
+
+	deinit {
+		PHPhotoLibrary.shared().unregisterChangeObserver(self)
+	}
+
+	// MARK: UIViewController life cycles
+
+	override func viewDidLoad() {
+		super.viewDidLoad()
+		self.collectionView.allowsMultipleSelection = true
+		self.collectionView?.contentInsetAdjustmentBehavior = .always
+
+		// Register collection view cell class
+		self.collectionView!.register(PhotoSelectionViewCell.self,
+									  forCellWithReuseIdentifier: PhotoSelectionViewCell.identifier)
+		thumbnailWidth = min(floor(self.view.bounds.size.width * thumbnailSizeMultiplier), thumbnailMaxWidth)
+		layout.itemSize = CGSize(width: thumbnailWidth, height: thumbnailWidth)
+
+		resetCachedAssets()
+
+		// Register observer to handle photo library changes
+		PHPhotoLibrary.shared().register(self)
+
+		// Set title for the navigation bar
+		if assetCollection != nil {
+			self.title = assetCollection!.localizedTitle
+			let fetchOptions = PHFetchOptions()
+			fetchOptions.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: false)]
+			fetchResult = PHAsset.fetchAssets(in: assetCollection!, options: fetchOptions)
+		} else {
+			self.title = "All Photos".localized
+		}
+
+		// If the fetchResult property was not pre-populdated, fetch all photos from the library
+		if fetchResult == nil {
+			let allPhotosOptions = PHFetchOptions()
+			allPhotosOptions.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: false)]
+			fetchResult = PHAsset.fetchAssets(with: allPhotosOptions)
+		}
+
+		self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
+
+		// Setup the toolbar buttons
+		let selectAllButtonItem = UIBarButtonItem(title: "Select All".localized, style: .done, target: self, action: #selector(selectAllItems))
+		let deselectAllButtonItem = UIBarButtonItem(title: "Deselect All".localized, style: .done, target: self, action: #selector(deselectAllItems))
+		let uploadButtonItem = UIBarButtonItem(title: "Upload".localized, style: .done, target: self, action: #selector(upload))
+		let flexibleSpaceButtonItem = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
+		self.toolbarItems = [selectAllButtonItem, flexibleSpaceButtonItem, uploadButtonItem, flexibleSpaceButtonItem, deselectAllButtonItem]
+
+		// Setup duration formatter
+		durationFormatter.unitsStyle = .positional
+		durationFormatter.allowedUnits = [.hour, .minute, .second]
+		durationFormatter.zeroFormattingBehavior = [.pad]
+	}
+
+	override func viewWillAppear(_ animated: Bool) {
+		super.viewWillAppear(animated)
+		Theme.shared.register(client: self)
+
+		self.navigationController?.isToolbarHidden = false
+
+		// Determine the size of the thumbnails to request from the PHCachingImageManager.
+		let scale = UIScreen.main.scale
+		if let cellSize = (self.collectionViewLayout as? UICollectionViewFlowLayout)?.itemSize {
```

Comment:
Maybe `cellSize` have to be calculated on `oriantationChanged`

Reply:
@hosy latest revision takes `orientation` into account